proposals/lip-0052.md

```diff
@@ -362,712 +379,1832 @@ The module provides the following commands to modify the NFT store.
+
+    if NFTStore(nftID) does not exist:
+        raise Exception("NFT substore entry does not exist.")
+
+    if len(getNFTOwner(nftID)) == CHAIN_ID_LENGTH:
+        raise Exception("NFT is escrowed to another chain.")
     else:
-        return entry does not exist
+        return userStore(getNFTOwner(nftID), nftID).lockingModule
 ```
 
-#### getNextAvailableIndex
+#### getNFTOwner
 
-This function returns the max index of a collection.
+This function returns the owner of an NFT.
 
 ```python
-getNextAvailableIndex(collection):
-    if collectionStore(collection) does not exist:
-        return collection does not exist
+def getNFTOwner(nftID: NFTID) -> Address:
 
-    return collectionStore(collection).nextAvailableIndex
+    if NFTStore(nftID) does not exist:
+        raise Exception("NFT substore entry does not exist.")
+
+    return NFTStore(nftID).owner
 ```
 
-#### getNextAvailableCollection
+#### getNextAvailableIndex
 
-This function returns the next available collection.
+This function returns the next available index of a collection.
 
 ```python
-getNextAvailableCollection():     
-    return nextAvailableCollection
+def getNextAvailableIndex(collectionID: CollectionID) -> uint64:
+
+    count = 0
+
+    for key in NFTStore.keys():
+        if key[`CHAIN_ID_LENGTH`:`CHAIN_ID_LENGTH`+`COLLECTION_ID_LENGTH`] == collectionID:
+            count +=1
+
+    return count
 ```
 
 #### create
 
 This function creates an NFT.
 
 ```python
-create(address, collection, attributes):
-    if (length(attributes) > MAX_BYTE_LENGTH_ATTRIBUTES bytes
-        or collectionStore(collection) does not exist):
-        create fails
-
-    index = collectionStore(collection).nextAvailableIndex
-    nftID = {"chainID": CHAIN_ID_ALIAS_NATIVE,
-             "collection": collection,
-             "index": index}
-    createNFTEntry(nftID, address, attributes)   
+def create(
+    address: Address,
+    collectionID: Collection
+) -> None:
+
+    index = getNextAvailableIndex(collectionID)
+    nftID = OWN_CHAIN_ID + collectionID + uint64be(index)
+
+    createNFTEntry(address, nftID)
     createUserEntry(address, nftID)
-    collectionStore(collection).nextAvailableIndex += 1
+
+    emitEvent(
+        module = MODULE_NAME_NFT,
+        name = EVENT_NAME_CREATE,
+        data = {
+            "address": address,
+            "nftID": nftID,
+            "collectionID": collectionID,
+            "result": RESULT_SUCCESSFUL
+        },
+        topics = [address, nftID]
+    )
 ```
 
 #### destroy
 
 This function destroys an NFT.
 
 ```python
-destroy(nftID):
-    nftID = canonicalNFTID(nftID)
-    if (nftID.chainID != CHAIN_ID_ALIAS_NATIVE
-        or NFTStore(nftID) does not exist):
-        destroy fails
-
-    address = NFTStore(nftID).owner
-    deleteNFTEntry(nftID)
-    deleteUserEntry(address, nftID)
+def destroy(
+    address: Address,
+    nftID: NFTID
+) -> None:
+
+    if getChainID(nftID) != OWN_CHAIN_ID:
```

Comment:
Is there a reason why we don't allow to destroy non-native NFTs?

Reply:
These attributes would be stored also in the native chain. In a sense, I think only the attributes stored in the native chain should be considered the 'real' ones.

Reply:
can be resolved